src/main/java/ca/uhn/fhir/jpa/starter/interceptor/CustomAuthorizationInterceptor.java

```diff
@@ -0,0 +1,178 @@
+
+	private final AppProperties config;
+
+	public CustomAuthorizationInterceptor(AppProperties config) {
+		super();
+		this.config = config;
+	}
+
+	@Override
+	public List<IAuthRule> buildRuleList(RequestDetails theRequest) {
+		if (!isAnyEnabled()) {
+			return authorizedRule();
+		}
+
+		try {
+			if (Constants.URL_TOKEN_METADATA.equals(theRequest.getRequestPath())) {
+				return unauthorizedRule();
+			}
+
+			if (isUsingOAuth(theRequest)) {
+				return authorizeOAuth(theRequest);
+			}
+			if (isUsingApiKey(theRequest)) {
+				return authorizeApiKey(theRequest);
+			}
+			logger.warn("Authorization failed - no authorization supplied");
+		} catch (AuthenticationException e) {
+			throw e;
+		} catch (RuntimeException e) {
+			logger.error("Unexpected exception during authorization: {}", e.getMessage());
+		}
+
+		throw new AuthenticationException("Missing or invalid authorization");
+	}
+
+	private boolean isAnyEnabled() {
+		return isOAuthEnabled() || isApiKeyEnabled();
+	}
+
+	private boolean isUsingOAuth(RequestDetails theRequest) {
+		return isOAuthEnabled() && OAuth2Helper.hasToken(theRequest);
+	}
+
+	private boolean isUsingApiKey(RequestDetails theRequest) {
+		return isApiKeyEnabled() && ApiKeyHelper.hasApiKey(theRequest);
+	}
+
+	private List<IAuthRule> authorizedRule() {
+		return new RuleBuilder()
+			.allowAll()
+			.build();
+	}
+
+	private List<IAuthRule> unauthorizedRule() {
+		// By default, deny everything except the metadata request
+		return new RuleBuilder()
+			.allow().metadata().andThen()
+			.denyAll()
+			.build();
+	}
+
+	private List<IAuthRule> authorizeOAuth(RequestDetails theRequest) throws AuthenticationException {
+		logger.info("Authorizing via OAuth2");
+		String token = OAuth2Helper.getToken(theRequest);
+		try {
+			DecodedJWT jwt = JWT.decode(token);
+			String jwksUrl = config.getOauth().getJwks_url();
+			OAuth2Helper.verify(jwt, jwksUrl);
+
+			List<String> clientRoles = OAuth2Helper.getClientRoles(jwt, getOAuthClientId());
+			if (clientRoles.isEmpty()) {
+				logger.warn("Authorization failure - token doesn't have any client roles");
+				return unauthorizedRule();
+			}
+
+			// The only difference between the admin role and the user role is that the admin role
+			// allows DELETE requests. It still needs to enforce a patient claim, if one exists.
+			if (theRequest.getRequestType().equals(RequestTypeEnum.DELETE)
+					&& !clientRoles.contains(getOAuthAdminRole())) {
+				logger.warn("Authorization failure - token doesn't have the admin role required for delete");
+				return unauthorizedRule();
+			}
+
+			if (clientRoles.contains(getOAuthAdminRole()) || clientRoles.contains(getOAuthUserRole())) {
+				String patientId = OAuth2Helper.getClaimAsString(jwt, "patient");
+				if (Strings.isNullOrEmpty(patientId)) {
+					logger.debug("No patient claim specified in authorization token");
+					return authorizedRule();
+				} else {
```

Comment:
Are we going to support delete operation for the patient claim token ? If yes then code needs to handle this.
Previously we were checking if the token has fhir4-admin role then we have authorizedRule() called.

Reply:
I'm not sure I understand what your concern is.

Previously, having the admin role would cause it to ignore the patient claim in the token. I consider that a bug. If the patient claim exists, it should be considered.

Reply:
I mean, are we going to support the delete operation for patient claim token which has fhir4-admin role as well ?
e.x. I have the fhir-admin role in auth server, I issued a patient token (specifying patient in header) and then try to delete some resources, in that case will it work ?

Reply:
Yes, we support that. Yes, it works with the code the way it is here.

Reply:
Can you confirm at your end, when I am deleting the patient using patient claim token, I am getting 403 access denied.

Reply:
I can confirm that it fails, but it returns a 404 if you try to delete a resource that the patient claim doesn't have access to. Not a 403. Again, returning a 403 "leaks" knowledge of a resource that the client doesn't have access to.

Reply:
@shubhamparikh927 I've pushed some updates which should address the deletion issue.